File: Code-CSharp/Base/Containers/Map/MapEntry.cs
using System.Collections.Immutable;

namespace CodeCsharp.Base.Containers.Map;

public class MapEntry {
  public static void MapProgram() {
    //initialize a map
    SortedDictionary<string, string> map = new SortedDictionary<string, string>();
    //initialize a map with value
    SortedDictionary<string, string> map1 = new SortedDictionary<string, string> {
        { "aa", "a" }, { "bb", "b" }, { "cc", "c" }, { "dd", "d" }
    };
    //copy map
    map = new SortedDictionary<string, string>(map1);
    //whether map is empty
    bool isEmpty = map.Count == 0;
    //clear map
    map.Clear();
    //insert element if not exist
    map.TryAdd("ee", "e");
    //insert or update element
    map["ee"] = "e2";
    //initialize a map contains values
    SortedDictionary<string, string> vec = new SortedDictionary<string, string> {
        { "aa", "a" }, { "bb", "b" }, { "cc", "c" }, { "dd", "d" }
    };
    //insert a range of elements if not exist
    foreach (var pair in vec) map.TryAdd(pair.Key, pair.Value);
    //insert or update a range of elements
    foreach (var pair in vec) map[pair.Key] = pair.Value;
    //get size of map
    int size = map.Count;
    //whether map contains element
    bool b1 = map.ContainsKey("bb");
    //create immutable reference of map
    ImmutableSortedDictionary<string, string> constMapRef = map.ToImmutableSortedDictionary();
    //whether two maps equal
    bool b2 = Enumerable.SequenceEqual(map, constMapRef);
    //remove element equal key
    map.Remove("cc");
    //get value of key or else default
    string defaultResult = map.GetValueOrDefault("cc", "default");
    //get optional value of key
    string? result = map.GetValueOrDefault("cc", null);
    //get optional value of first key lower or equal given key
    result = map.FirstOrDefault(e => string.CompareOrdinal(e.Key, "cc") <= 0,
        new KeyValuePair<string, string>(string.Empty, null)).Value;
    //get optional value of first key greater or equal given key
    result = map.LastOrDefault(e => string.CompareOrdinal(e.Key, "cc") >= 0,
        new KeyValuePair<string, string>(string.Empty, null)).Value;
  }

  public static bool TestAll() {
    bool ret = true;
    SortedDictionary<int, int> c, c1, c2, c3, c4, c5, c6, c7, c8;
    SortedDictionary<int, int> m = new SortedDictionary<int, int> {
        { 1, 10 }, { 2, 8 }, { 3, 13 }, { 4, 9 }
    };
    c = new SortedDictionary<int, int> { { 1, 10 }, { 2, 8 }, { 3, 13 }, { 4, 9 } };
    c.Clear();
    c1 = new SortedDictionary<int, int>(c);
    c = new SortedDictionary<int, int>(m);
    c2 = new SortedDictionary<int, int>(c);
    c.TryAdd(2, 10);
    c3 = new SortedDictionary<int, int>(c);
    c.TryAdd(5, 10);
    c4 = new SortedDictionary<int, int>(c);
    c[2] = 10;
    c5 = new SortedDictionary<int, int>(c);
    m = new SortedDictionary<int, int> { { 0, 3 }, { 1, 8 }, { 2, 7 }, { 10, 3 } };
    foreach (var pair in m) c.TryAdd(pair.Key, pair.Value);
    c6 = new SortedDictionary<int, int>(c);
    foreach (var pair in m) c[pair.Key] = pair.Value;
    c7 = new SortedDictionary<int, int>(c);
    c.Remove(2);
    c8 = new SortedDictionary<int, int>(c);
    ret = c1.Count == 0
          && c.ContainsKey(3)
          && c.GetValueOrDefault(3) == 13
          && c.First(e => e.Key >= 2).Value == 13
          && c.Last(e => e.Key <= 2).Value == 8
        && c2.SequenceEqual(
            new SortedDictionary<int, int> { { 1, 10 }, { 2, 8 }, { 3, 13 }, { 4, 9 } })
        && c3.SequenceEqual(
            new SortedDictionary<int, int> { { 1, 10 }, { 2, 8 }, { 3, 13 }, { 4, 9 } })
        && c4.SequenceEqual(new SortedDictionary<int, int> {
            { 1, 10 }, { 2, 8 }, { 3, 13 }, { 4, 9 }, { 5, 10 }
        })
        && c5.SequenceEqual(new SortedDictionary<int, int> {
            { 1, 10 }, { 2, 10 }, { 3, 13 }, { 4, 9 }, { 5, 10 }
        })
        && c6.SequenceEqual(new SortedDictionary<int, int> {
            { 0, 3 }, { 1, 10 }, { 2, 10 }, { 3, 13 }, { 4, 9 }, { 5, 10 }, { 10, 3 }
        })
        && c7.SequenceEqual(new SortedDictionary<int, int> {
            { 0, 3 }, { 1, 8 }, { 2, 7 }, { 3, 13 }, { 4, 9 }, { 5, 10 }, { 10, 3 }
        })
        && c8.SequenceEqual(new SortedDictionary<int, int> {
            { 0, 3 }, { 1, 8 }, { 3, 13 }, { 4, 9 }, { 5, 10 }, { 10, 3 }
        });
    if (!ret)
      Console.WriteLine("Map Test Failed!");
    return ret;
  }
}

File: Code-Cpp/code_cpp/base/containers/map/map_entry.cc
#include "map_entry.h"
#include <vector>
#include <optional>

void MapEntry::MapProgram() {
  //initialize a map
  std::map<std::string, std::string> map;
  //initialize a map with value
  std::map<std::string, std::string> map1{{"aa", "a"}, {"bb", "b"}, {"cc", "c"}, {"dd", "d"}};
  //copy map
  map = map1;
  //whether map is empty
  bool is_empty = map.empty();
  //clear map
  map.clear();
  //insert element if not exist
  map.try_emplace("ee", "e");
  //insert or update element
  map["ee"] = "e2";
  //initialize a map contains values
  std::map<std::string, std::string> vec{{"aa", "a"}, {"bb", "b"}, {"cc", "c"}, {"dd", "d"}};
  //insert a range of elements if not exist
  map.merge(vec);
  //insert or update a range of elements
  for (const auto &p : vec)
    map[p.first] = p.second;
  //get size of map
  int size = map.size();
  //whether map contains element
  bool b1 = map.count("bb") > 0; // map.contains("bb");
  //create immutable reference of map
  const std::map<std::string, std::string> &kConstMapRef = map;
  //whether two maps equal
  bool b2 = map == kConstMapRef;
  //remove element equal key
  map.erase("cc");
  //get value of key or else default
  auto it = map.find("cc");
  std::string default_result = it != map.end() ? it->second : "default";
  //get optional value of key
  std::optional<std::string> result = it != map.end() ? std::make_optional(it->second) : std::nullopt;
  //get optional value of first key lower or equal given key
  it = --map.upper_bound("cc");
  result = it != map.end() ? std::make_optional(it->second) : std::nullopt;
  //get optional value of first key greater or equal given key
  it = map.lower_bound("cc");
  result = it != map.end() ? std::make_optional(it->second) : std::nullopt;
}
bool MapEntry::TestAll() {
  bool ret = true;
  using namespace std;
  map<int, int> c, c1, c2, c3, c4, c5, c6, c7, c8;
  map<int, int> m{{1, 10}, {2, 8}, {3, 13}, {4, 9}};
  c = {{1, 10}, {2, 8}, {3, 13}, {4, 9}};
  c.clear();
  c1 = c;
  c = m;
  c2 = c;
  c.try_emplace(2, 10);
  c3 = c;
  c.try_emplace(5, 10);
  c4 = c;
  c[2] = 10;
  c5 = c;
  m = {{0, 3}, {1, 8}, {2, 7}, {10, 3}};
  c.merge(m);
  c6 = c;
  for (const auto &p : m)
    c[p.first] = p.second;
  c7 = c;
  c.erase(2);
  c8 = c;
  ret = c1.empty()
      && c.count(3) > 0
      && c.find(3)->second == 13
      && c.lower_bound(2)->second == 13
      && (--c.upper_bound(2))->second == 8
      && c2 == map<int, int>{{1, 10}, {2, 8}, {3, 13}, {4, 9}}
      && c3 == map<int, int>{{1, 10}, {2, 8}, {3, 13}, {4, 9}}
      && c4 == map<int, int>{{1, 10}, {2, 8}, {3, 13}, {4, 9}, {5, 10}}
      && c5 == map<int, int>{{1, 10}, {2, 10}, {3, 13}, {4, 9}, {5, 10}}
      && c6 == map<int, int>{{0, 3}, {1, 10}, {2, 10}, {3, 13}, {4, 9}, {5, 10}, {10, 3}}
      && c7 == map<int, int>{{0, 3}, {1, 8}, {2, 7}, {3, 13}, {4, 9}, {5, 10}, {10, 3}}
      && c8 == map<int, int>{{0, 3}, {1, 8}, {3, 13}, {4, 9}, {5, 10}, {10, 3}};
  if (!ret)
    cout << "Map Test Failed!"<< endl;
  return ret;
}
